src/appleseed/foundation/image/genericimagefilewriter.cpp

```diff
@@ -31,50 +30,481 @@
 #include "genericimagefilewriter.h"
 
 // appleseed.foundation headers.
-#include "foundation/core/exceptions/exceptionunsupportedfileformat.h"
-#include "foundation/image/oiioimagefilewriter.h"
+#include "foundation/core/exceptions/exceptionioerror.h"
+#include "foundation/image/icanvas.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/compiler.h"
+#include "foundation/utility/foreach.h"
+#include "foundation/utility/iostreamop.h"
 #include "foundation/utility/string.h"
 
 // Boost headers.
 #include "boost/filesystem/path.hpp"
 
 // Standard headers.
-#include <string>
-
-using namespace std;
-namespace bf = boost::filesystem;
+#include <algorithm>
+#include <cstring>
+#include <memory>
+#include <stdexcept>
+#include <vector>
 
 namespace foundation
 {
 
-//
-// GenericImageFileWriter class implementation.
-//
+struct GenericImageFileWriter::Impl
+{
+    std::vector<const ICanvas*>     m_canvas;
+    std::vector<OIIO::ImageSpec>    m_spec;
+};
+
+GenericImageFileWriter::GenericImageFileWriter(const char* filename) :
+    impl( new Impl() )
+{
+    assert(filename);
+
+    m_filename = filename;
+
+    m_writer = OIIO::ImageOutput::create(m_filename);
+    if (m_writer == nullptr)
+    {
+        const std::string msg = OpenImageIO::geterror();
+        throw ExceptionIOError(msg.c_str());
+    }
+}
+
+GenericImageFileWriter::~GenericImageFileWriter()
+{
+    // Destroy the ImageOutput stucture.
+    OIIO::ImageOutput::destroy(m_writer);
+
+    delete impl;
+}
+
+size_t GenericImageFileWriter::get_image_count() const
+{
+    return impl->m_canvas.size();
+}
 
-void GenericImageFileWriter::write(
-    const char*             filename,
-    const ICanvas&          image,
-    const ImageAttributes&  image_attributes)
+void GenericImageFileWriter::append_image(const ICanvas* image)
 {
-    const bf::path filepath(filename);
-    const string extension = lower_case(filepath.extension().string());
+    assert(image);
 
-    if (extension == ".exr" || extension == ".png")
+    impl->m_canvas.push_back(image);
+    impl->m_spec.push_back(OIIO::ImageSpec{});
+
+    set_image_spec();
+}
+
+OIIO::TypeDesc convert_pixel_format(const PixelFormat format)
```

Comment:
Wrap local helper functions in anonymous namespace:
```cpp
namespace
{
    OIIO::TypeDesc convert_pixel_format(const PixelFormat format)
    {
        ...
    }
}
```

Reply:
Done